include/power/pm_policy.h

```diff
@@ -0,0 +1,183 @@
+ * @param policy_name The name of this policy
+ * @param policy_next_state_get Callback to get the next PM state
+ * @param policy_low_power_devices_get Callback to determine whether to put
+ *        devices in low power state, given the system PM state
+ */
+#define PM_POLICY_DEFINE(policy_type, policy_name, policy_next_state_get, \
+			 policy_low_power_devices_get, ...)		\
+	Z_STRUCT_SECTION_ITERABLE_ALTERNATE(pm_policy, policy_type,	\
+					    policy_name) = {		\
+		.name = STRINGIFY(policy_name),			\
+		.len = sizeof(struct policy_type),			\
+		.next_state_get = policy_next_state_get,		\
+		.low_power_devices_get = policy_low_power_devices_get,	\
+		__VA_ARGS__						\
+	}
+
+#define PM_POLICY_FOREACH(struct_type, iterator)			\
+	extern struct struct_type _CONCAT(_##struct_type, _list_start)[]; \
+	extern struct struct_type _CONCAT(_##struct_type, _list_end)[]; \
+	for (struct struct_type *iterator =				\
+		     _CONCAT(_##struct_type, _list_start);		\
+	     ({ __ASSERT(iterator <= _CONCAT(_##struct_type, _list_end), \
+			 "unexpected list end location");		\
+		     iterator < _CONCAT(_##struct_type, _list_end); }); \
+	     iterator = (struct struct_type *)				\
+		     ((uint8_t *)iterator + iterator->len))
+
+/**
+ * @brief Function to create device PM list
+ */
+void sys_pm_create_device_list(void);
+
+/**
+ * @brief Function to suspend the devices in PM device list
+ */
+int sys_pm_suspend_devices(void);
+
+/**
+ * @brief Function to put the devices in PM device list in low power state
+ */
+int sys_pm_low_power_devices(void);
+
+/**
+ * @brief Function to force suspend the devices in PM device list
+ */
+int sys_pm_force_suspend_devices(void);
+
+/**
+ * @brief Function to resume the devices in PM device list
+ */
+void sys_pm_resume_devices(void);
+
+/**
+ * @brief Function to get the next PM state based on the ticks
+ */
+static inline enum power_states sys_pm_policy_next_state(int32_t ticks)
+{
+	extern struct pm_policy *sys_pm_policy;
+
+	if (sys_pm_policy == NULL) {
+		return -EINVAL;
+	}
+
+	if (sys_pm_policy->next_state_get == NULL) {
+		/* Should it be -ENOSYS ? */
+		return -ENOTSUP;
+	}
+
+	return sys_pm_policy->next_state_get(sys_pm_policy, ticks);
+}
+
+/**
+ * @brief Function to determine whether to put devices in low
+ *        power state, given the system PM state.
+ */
+static inline bool sys_pm_policy_low_power_devices(enum power_states state)
+{
+	extern struct pm_policy *sys_pm_policy;
+
+	if (sys_pm_policy == NULL) {
+		return -EINVAL;
+	}
+
+	if (sys_pm_policy->low_power_devices_get == NULL) {
+		/* Should it be -ENOSYS ? */
+		return -ENOTSUP;
+	}
+
+	return sys_pm_policy->low_power_devices_get(sys_pm_policy, state);
+}
+
+/**
+ * @brief Function to set the system power management policy
+ *
+ * @param policy_name The policy name to be used. @see PM_POLICY_DEFINE
+ *
+ * @return false if the policy is not available, true otherwise.
+ */
+static inline bool sys_pm_policy_set(const char *policy_name)
```

Comment:
What will be invoking this function?  Is it reasonable to do it by name, rather than passing a reference to the policy structure the application wants to use?

Reply:
This is just more convenient for the application, instead of having to do:
```
extern struct pm_policy *dummy_policy;
sys_pm_policy_set(dummy_policy);
```

Reply:
Why would the policy be extern?  Isn't the app providing the policy, so it'd not be an external global symbol, but a static local one?  If there are default policies available to reference they should be declared in a header somewhere.

Reply:
The idea is that the symbol is implicitly created by the macro, and can be in a different object file, not necessarily in the same  file that the application will configure the policy, this is the same behavior of other similar macros.

Though as we had discussed in a previous meeting, there is no support for multiple policies  anymore, so this is not relevant anymore.